daemon/oci_linux.go

```diff
@@ -98,6 +100,30 @@ func setDevices(s *specs.Spec, c *container.Container) error {
 		for _, d := range hostDevices {
 			devs = append(devs, oci.Device(d))
 		}
+		for _, deviceMapping := range c.HostConfig.Devices {
+			if deviceMapping.PathOnHost == deviceMapping.PathInContainer {
+				if deviceMapping.CgroupPermissions != "rwm" {
+					logrus.WithField("container", c.ID).Warnf("custom %s permissions for device %s are ignored in privileged mode", deviceMapping.CgroupPermissions, deviceMapping.PathOnHost)
+				}
+				continue
+			}
+			var err error
+			if _, err = os.Stat(deviceMapping.PathInContainer); os.IsNotExist(err) {
+				if deviceMapping.CgroupPermissions != "rwm" {
+					logrus.WithField("container", c.ID).Warnf("custom %s permissions for device %s are ignored in privileged mode", deviceMapping.CgroupPermissions, deviceMapping.PathOnHost)
+				}
+				d, _, error := oci.DevicesFromPath(deviceMapping.PathOnHost, deviceMapping.PathInContainer, "rwm")
+				if error != nil {
+					return error
+				}
+				devs = append(devs, d...)
+			} else {
```

Comment:
Was looking if we could get rid old all these nested ifs (for example, but extracting the "non-privileged" and "privileged" branches to separate functions), but may be best to leave that refactoring for another PR